pmonitor.1
.TH PMONITOR 1 "13 February 2018"
.\"
.\" (C) Copyright 2006-2018 Diomidis Spinellis
.\"
.\"  Licensed under the Apache License, Version 2.0 (the "License");
.\"  you may not use this file except in compliance with the License.
.\"  You may obtain a copy of the License at
.\"
.\"      http://www.apache.org/licenses/LICENSE-2.0
.\"
.\"  Unless required by applicable law or agreed to in writing, software
.\"  distributed under the License is distributed on an "AS IS" BASIS,
.\"  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
.\"  See the License for the specific language governing permissions and
.\"  limitations under the License.
.\"
.SH NAME
pmonitor \- monitor a job's progress
.SH SYNOPSIS
\fBpmonitor\fP
\fB\-c\fP \fIcommand\fP |
\fB\-f\fP \fIfile\fP |
\fB\-p\fP \fIpid\fP
[\fB-i\fP \fIinterval\fP]
[\fB-d\fP]
[\fB-u\fP]
.SH DESCRIPTION
The \fIpmonitor\fP command will display the progress of a process
as a percentage.
It does this by examining the process's open files,
and calculating the ratio between
the position of the file's seek pointer offset and the file length.
For processes that process files in a sequential fashion,
such as file compression and database import, this ratio
can be translated to the percentage of the job that has been completed.
.PP
The command may produce no output, if a process does not hold any open
files with a non-zero seek offset.
.SH OPTIONS
.TP
\fB\-c\fP, \fB\-\-command\fP=\fICOMMAND\fP
Monitor the progress through the files opened by the specified running command.
The command can be specified through (part) of its name
(e.g. \fIgzip\fP), or through a regular expression if the
argument starts with a \fC/\fP.
See the documentation of the \fIlsof\fP(1) \fB\-c\fP option regarding
the rules and flags associated with the use of regular expressions.

.TP
\fB\-d\fP, \fB\-\-diff\fP
In repeated output mode,
output only records that differ from one iteration to the next.

.TP
\fB\-f\fP, \fB\-\-file\fP=\fIFILE\fP
Monitor a process's progress through its processing of the specified file.

.TP
\fB\-p\fP, \fB\-\-pid\fP=\fIPID\fP
Monitor the progress of the process with the specified process id \fIPID\fP.

.TP
\fB\-i\fP, \fB\-\-interval\fP=\fIINTERVAL\fP
Repeat the output every \fIINTERVAL\fP seconds.

.TP
\fB\-h\fP, \fB\-\-help\fP
Display the program's usage information and exit.

.TP
\fB\-u\fP, \fB\-\-update\fP
Also monitor files that are opened in update (read and write) mode,
rather than only those that are opened in read-only mode.
This is useful for monitoring table scans of RDBMS engines.

.SH AUTHOR
Diomidis Spinellis \(em <http://www.spinellis.gr>

.SH "REPORTING BUGS"
Visit the utility's \fIGitHub\fP page at
<https://github.com/dspinellis/pmonitor>.
Fixes and improvements are accepted through pull requests.

.SH "SEE ALSO"
\fIlsof\fP(1),
.PP
The \fIpmonitor\fP command is modelled after a similar facility
that was available on the Permin-Elmer/Concurrent OS32.


pmonitor.sh
#!/bin/sh
#
# Monitor the progress of a specified job
#
# Copyright 2006-2018 Diomidis Spinellis
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.
#
# For each file or file associated with the specified process is reading,
# display the percentage associated with its seek pointer offset.  For
# files that are processed in a sequential fashion this can be translated
# to the percentage of the job that has been completed.
#
# This command is modelled after a similar facility
# available on Permin-Elmer/Concurrent OS32
#
# Requires:
# - lsof(8) with offset (-o) printing functionality
#
# Submit issues and pull requests as https://github.com/dspinellis/pmonitor
#

# Run lsof with the specified options
# The OPT1 and OPT2 variables are passed to lsof as arguments.
opt_lsof()
{
  lsof -w -o0 -o "$OPT1" "$OPT2"
}

# Display the scanned percentage of lsof files.
display()
{
  # Obtain the offset and print it as a percentage
  awk '
    # Return current time
    function time() {
	"date +%s" | getline t
	close("date +%s")
	return t
    }

    # Return length of specified file
    function file_length(fname) {
      if (!cached_length[fname]) {
        if (fname ~ /^\/dev\/[^/]*$/ && system("test -b " fname) == 0) {
          getline < ("/sys/block/" substr(fname, 6) "/size")
          cached_length[fname] = $1 * 512 # sector size is always 512 bytes
        } else {
          "ls -l '\''" fname "'\'' 2>/dev/null" | getline
          cached_length[fname] = $5 + 0
        }
      }
      return cached_length[fname]
    }

    BEGIN {
      CONVFMT = "%.2f"
      start = time()
    }

    $4 ~ /^[0-9]+[r'$UPDATE']$/ && $7 ~ /^0t/ {
      now = time()
      offset = substr($7, 3)
      fname = $9
      len = file_length(fname)
      if (len > 0) {
	if (!start_offset[fname])
	  start_offset[fname] = offset
	delta_t = now - start
	delta_o = offset - start_offset[fname]
	if (delta_t > 2 && delta_o > 0) {
	  bps = delta_o / delta_t
	  t = (len - offset) / bps
	  eta = ""
	  if (t > 0) {
	    eta_s = t % 60
	    t = int(t / 60)
	    eta_m = t % 60
	    t = int(t / 60)
	    eta_h = t
	    eta = sprintf(" ETA %d:%02d:%02d", eta_h, eta_m, eta_s)
	  }
	}
        out = fname "\t" offset / len * 100 "%"
	if (!'$ONLYDIFF' || !seen[out])
	  print out eta
	seen[out] = 1
      }
    }
  '
}

# Report program usage information
usage()
{
	cat <<\EOF 1>&2
Usage:

pmonitor [-c command] [-d] [-f file] [-i interval] [-p pid]
-c, --command=COMMAND	Monitor the progress of the specified running command
-d, --diff		During continuous display show only records that differ
-f, --file=FILE		Monitor the progress of commands processing the
			specified file
-h, --help		Display this message and exit
-i, --interval=INTERVAL	Continuously display the progress every INTERVAL seconds
-p, --pid=PID		Monitor the progress of the process with the specified
			process id
-u, --update		Also monitor files opened in update (rather than read
			mode)

Exactly one of the c, f, p options must be specified.

Terminating...
EOF
}

# Option processing; see /usr/share/doc/util-linux-ng-2.17.2/getopt-parse.bash
# Note that we use `"$@"' to let each command-line parameter expand to a
# separate word. The quotes around `$@' are essential!
# We need TEMP as the `eval set --' would nuke the return value of getopt.

# Allowed short options
SHORTOPT=c:,d,f:,h,i:,p:,u

if getopt -l >/dev/null 2>&1 ; then
  # Simple (e.g. FreeBSD) getopt
  TEMP=$(getopt $SHORTOPT "$@")
else
  # Long options supported
  TEMP=$(getopt -o $SHORTOPT --long command:,diff,file:,help,interval:,pid:,update -n 'pmonitor' -- "$@")
fi

if [ $? != 0 ] ; then
  usage
  exit 2
fi

# Note the quotes around `$TEMP': they are essential!
eval set -- "$TEMP"

ONLYDIFF=0

while : ; do
  case "$1" in
    -c|--command)
      test -z "$OPT1" || usage
      OPT1=-c
      OPT2="$2"
      shift 2
      ;;
    -d|--diff)
      ONLYDIFF=1
      shift
      ;;
    -f|--file)
      test -z "$OPT1" || usage
      if ! [ -r "$2" ] ; then
	echo "Unable to read $2" 1>&2
	exit 1
      fi
      OPT1=--
      OPT2="$2"
      shift 2
      ;;
    -i|--interval)
      INTERVAL="$2"
      shift 2
      ;;
    -h|--help)
      usage
      exit 0
      ;;
    -p|--pid)
      test -z "$OPT1" || usage
      OPT1=-p
      OPT2="$2"
      shift 2
      ;;
    -u|--update)
      UPDATE=u
      shift
      ;;
    --)
      shift
      break
      ;;
    *)
      echo "Internal error!"
      exit 3
      ;;
  esac
done

# No more arguments allowed and one option must be specified
if [ "$1" != '' -o ! -n "$OPT1" -o ! -n "$OPT2" ]
then
  usage
  exit 2
fi

if [ "$INTERVAL" ] ; then
  while : ; do
    opt_lsof
    sleep $INTERVAL
  done
else
  ONLYDIFF=0
  opt_lsof
fi |
display
